concordium-consensus/src/Concordium/External/DryRun.hs

```diff
@@ -0,0 +1,881 @@
+import Control.Monad.Trans.Cont (ContT (..))
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Short as SBS
+import qualified Data.ByteString.Unsafe as BS
+import Data.IORef
+import qualified Data.Map.Strict as Map
+import qualified Data.ProtoLens as Proto
+import Foreign
+import GHC.Stack
+import Lens.Micro.Platform
+
+import qualified Concordium.Crypto.SignatureScheme as Sig
+import Concordium.GRPC2
+import Concordium.Logger
+import Concordium.Types
+import Concordium.Types.Execution
+import qualified Concordium.Types.InvokeContract as InvokeContract
+import Concordium.Types.Parameters
+import Concordium.Types.Queries
+import Concordium.Types.Transactions
+import Concordium.Utils
+import qualified Proto.V2.Concordium.Types as Proto
+
+import qualified Concordium.Cost as Cost
+import qualified Concordium.External as Ext
+import Concordium.External.Helpers
+import Concordium.GlobalState.BlockMonads
+import Concordium.GlobalState.BlockState
+import Concordium.GlobalState.Persistent.BlockState
+import qualified Concordium.GlobalState.TreeState as SkovV0
+import Concordium.GlobalState.Types (BlockStateTypes (..))
+import Concordium.ID.Types
+import qualified Concordium.KonsensusV1.SkovMonad as SkovV1
+import qualified Concordium.KonsensusV1.Types as SkovV1
+import qualified Concordium.Kontrol as SkovV0
+import Concordium.MultiVersion
+import Concordium.Queries
+import qualified Concordium.Scheduler as Scheduler
+import qualified Concordium.Scheduler.Environment as Scheduler
+import qualified Concordium.Scheduler.EnvironmentImplementation as Scheduler
+import qualified Concordium.Scheduler.InvokeContract as InvokeContract
+import qualified Concordium.Skov as SkovV0
+import qualified Concordium.TransactionVerification as TVer
+
+-- | The charge to the energy quota for loading a block state.
+costLoadBlockState :: Energy
+costLoadBlockState = 2000
+
+-- | The charge to the energy quota for getting account info.
+costGetAccountInfo :: Energy
+costGetAccountInfo = 200
+
+-- | The charge to the energy quota for getting contract instance info.
+costGetInstanceInfo :: Energy
+costGetInstanceInfo = 200
+
+-- | The base charge to the energy quota for running invoke instance on a contract.
+--  In addition, the energy cost of the execution will be charged.
+costInvokeInstanceBase :: Energy
+costInvokeInstanceBase = 200
+
+-- | The charge to the energy quota for setting the block timestamp.
+costSetTimestamp :: Energy
+costSetTimestamp = 50
+
+-- | The charge to the energy quota for minting to an account.
+costMintToAccount :: Energy
+costMintToAccount = 400
+
+-- | The base cost for dry-running a transaction.
+--  In addition, the energy cost of the execution will be charged.
+costTransactionBase :: Energy
+costTransactionBase = 400
+
+-- | An opaque rust vector.
+data ForeignVec
+
+-- | A function that copies bytes to a rust vector.
+type CopyToForeignVec = Ptr ForeignVec -> Ptr Word8 -> Int64 -> IO ()
```

Comment:
Don't we have this already in External? We certainly do this for all streaming calls.